src/chainstate/stacks/index/bits.rs

```diff
@@ -0,0 +1,419 @@
+#[inline]
+pub fn get_nodetype_hash(node: &TrieNodeType, child_hashes: &Vec<TrieHash>) -> TrieHash {
+    match node {
+        TrieNodeType::Leaf(ref data) => get_node_hash(data, child_hashes),
+        TrieNodeType::Node4(ref data) => get_node_hash(data, child_hashes),
+        TrieNodeType::Node16(ref data) => get_node_hash(data, child_hashes),
+        TrieNodeType::Node48(ref data) => get_node_hash(data, child_hashes),
+        TrieNodeType::Node256(ref data) => get_node_hash(data, child_hashes)
+    }
+}
+
+/// Calculate the hash of a TrieNode, given a byte buffer encoding all of its children's hashes.
+pub fn get_node_hash_bytes<T: TrieNode + std::fmt::Debug>(node: &T, child_hash_bytes: &Vec<u8>) -> TrieHash {
+    assert_eq!(child_hash_bytes.len() % TRIEHASH_ENCODED_SIZE, 0);
+
+    let mut hasher = TrieHasher::new();
+    let mut buf = Vec::with_capacity(node.byte_len());
+
+    node.to_consensus_bytes(&mut buf);
+    hasher.input(&buf);
+    hasher.input(child_hash_bytes);
+    
+    let mut res = [0u8; 32];
+    res.copy_from_slice(hasher.result().as_slice());
+
+    let ret = TrieHash(res);
+
+    if is_trace() {
+        // not in prod -- can spend a few cycles on fancy debug output
+        if child_hash_bytes.len() >= 50 {
+            // extract individual hashes
+            let mut all_hashes = Vec::with_capacity(child_hash_bytes.len() / TRIEHASH_ENCODED_SIZE);
+            for i in 0..child_hash_bytes.len() / TRIEHASH_ENCODED_SIZE {
+                let mut h_slice = [0u8; TRIEHASH_ENCODED_SIZE];
+                h_slice.copy_from_slice(&child_hash_bytes[TRIEHASH_ENCODED_SIZE*i..TRIEHASH_ENCODED_SIZE*(i+1)]);
+                all_hashes.push(TrieHash(h_slice))
+            }
+            trace!("get_node_hash_bytes: hash {:?} = {:?} + {:?}... ({})", &ret, node, &all_hashes, child_hash_bytes.len());
+        }
+        else {
+            trace!("get_node_hash_bytes: hash {:?} = {:?} + {:?}... ({})", &ret, node, &child_hash_bytes, child_hash_bytes.len());
+        }
+    }
+    ret
+}
+
+#[inline]
+pub fn get_nodetype_hash_bytes(node: &TrieNodeType, child_hash_bytes: &Vec<u8>) -> TrieHash {
+    match node {
+        TrieNodeType::Node4(ref data) => get_node_hash_bytes(data, child_hash_bytes),
+        TrieNodeType::Node16(ref data) => get_node_hash_bytes(data, child_hash_bytes),
+        TrieNodeType::Node48(ref data) => get_node_hash_bytes(data, child_hash_bytes),
+        TrieNodeType::Node256(ref data) => get_node_hash_bytes(data, child_hash_bytes),
+        TrieNodeType::Leaf(ref data) => get_node_hash_bytes(data, child_hash_bytes),
+    }
+}
+
+
+/// Low-level method for reading a TrieHash into a byte buffer from a Read-able and Seek-able struct.
+/// The byte buffer must have sufficient space to hold the hash, or this program panics.
+pub fn read_hash_bytes<F: Read + Seek>(f: &mut F, buf: &mut Vec<u8>) -> Result<(), Error> {
+    let mut hashbytes = [0u8; 32];
+    f.read(&mut hashbytes)
```

Comment:
`read` isn't guaranteed to fill its buffer, `read_exact` would be though

Reply:
Good catch!  Thanks!